tests/check_eventloop_udp.c

```diff
@@ -0,0 +1,385 @@
+
+        /* open a reader connection */
+        UA_UInt16 port = 30000;
+        UA_Variant portVar;
+        UA_Variant_setScalar(&portVar, &port, &UA_TYPES[UA_TYPES_UINT16]);
+        UA_ConnectionManager *cmReader = UA_ConnectionManager_new_POSIX_UDP(UA_STRING("udpCM"));
+        cmReader->connectionCallback = connectionCallback;
+        UA_KeyValueMap_set(&cmReader->eventSource.params,
+                           &cmReader->eventSource.paramsSize,
+                           UA_QUALIFIEDNAME(0, "listen-port"), &portVar);
+        el->registerEventSource(el, &cmReader->eventSource);
+
+        connCount = 0;
+        el->start(el);
+
+        /* Open a writer connection */
+        /* create writer eventloop */
+        UA_EventLoop *elWrite = UA_EventLoop_new_POSIX(UA_Log_Stdout);
+        clientId = 0;
+
+        UA_String targetHost = UA_STRING("localhost");
+        UA_KeyValuePair params[2];
+        params[0].key = UA_QUALIFIEDNAME(0, "port");
+        params[0].value = portVar;
+        params[1].key = UA_QUALIFIEDNAME(0, "hostname");
+        UA_Variant_setScalar(&params[1].value, &targetHost, &UA_TYPES[UA_TYPES_STRING]);
+
+        UA_ConnectionManager *cmWriter = UA_ConnectionManager_new_POSIX_UDP(UA_STRING("udpCM"));
+
+        /* the cm writer should not need a connection callback */
+        cmWriter->connectionCallback = connectionCallback;
+        elWrite->registerEventSource(elWrite, &cmWriter->eventSource);
+        elWrite->start(elWrite);
+        UA_StatusCode retval = cmWriter->openConnection(cmWriter, 2, params, (void*)0x01);
+
+        ck_assert_uint_eq(retval, UA_STATUSCODE_GOOD);
+        for(size_t i = 0; i < 2; i++) {
+            UA_DateTime next = el->run(el, 1);
+            UA_fakeSleep((UA_UInt32)((next - UA_DateTime_now()) / UA_DATETIME_MSEC));
+        }
+        /* The reader el should not generate any connections */
+        ck_assert_uint_eq(connCount, 0);
+
+        /* The writer el should receive a signal "ready to be written on" */
+        ck_assert_uint_eq(retval, UA_STATUSCODE_GOOD);
+        for(size_t i = 0; i < 2; i++) {
+            UA_DateTime next = elWrite->run(elWrite, 1);
+            UA_fakeSleep((UA_UInt32)((next - UA_DateTime_now()) / UA_DATETIME_MSEC));
+        }
+        ck_assert_uint_eq(connCount, 1);
+
+        /* Send a message from the writer */
+        received = false;
+        UA_ByteString snd;
+        retval = cmWriter->allocNetworkBuffer(cmWriter, clientId, &snd, strlen(testMsg));
+        ck_assert_uint_eq(retval, UA_STATUSCODE_GOOD);
+        memcpy(snd.data, testMsg, strlen(testMsg));
+        retval = cmWriter->sendWithConnection(cmWriter, clientId, 0, NULL, &snd);
+        ck_assert_uint_eq(retval, UA_STATUSCODE_GOOD);
+        for(size_t i = 0; i < 2; i++) {
+            UA_DateTime next = el->run(el, 1);
+            UA_fakeSleep((UA_UInt32)((next - UA_DateTime_now()) / UA_DATETIME_MSEC));
+        }
+        ck_assert(received);
```

Comment:
´el->run` iterates the eventloop after `sendWithConnection` and calls the defined `connectionCallback` where the `testMsg` content is asserted. `ck_assert(received)` just then asserts the state that was changed in the `connectionCallback` for the `received` condition.